src/axolotl/utils/models.py

```diff
@@ -446,42 +502,9 @@ def load_model(
             max_memory[i] = gpu_memory_limit
         max_memory["cpu"] = "256GiB"  # something sufficiently large to fit anything
 
-    if max_memory is not None:
-        # Based on https://github.com/togethercomputer/OpenChatKit/blob/main/inference/bot.py
-        from accelerate import infer_auto_device_map
-
-        with init_empty_weights():
-            model_canvas = AutoModelForCausalLM.from_config(
-                model_config, trust_remote_code=cfg.trust_remote_code or False
-            )
-        model_canvas.tie_weights()
-        device_map = infer_auto_device_map(
-            model_canvas,
-            max_memory=max_memory,
-            dtype=cfg.torch_dtype,
-        )
-        # We can discard max_memory now as we have a device map set up for us
-        max_memory = None
-
-    model_kwargs["device_map"] = device_map
+    set_model_device(cfg, max_memory, model_config, model_kwargs, device_map)
```

Comment:
the way python passes by reference and updates them in the function feels a bit awkward here. Not sure right now what a good solution would be to make this more obvious.

Reply:
I think the simplest way is **making `model_kwargs` as the return value of func `set_model_device`**. Simple but the obvious effect may be similar to the current one. And `model_kwargs` itself is mutable, this way probably make little sense.

A more complicated solution would be write a **class `ModelKwargs`**, with member functions like `__init__`, `update_model_device`, `update_dtype`, `update_attention`, `update_quantization` ... These funcs will be called in `load_model`, **making `load_model` and the change of `model_kwargs` more clearly**. However, this would bring **a lot changes** into `src/axolotl/utils/models.py` and **may exsisting some issues**, time is needed for validating it.